apps/cfp/event_tickets_lottery.py

```diff
@@ -0,0 +1,126 @@
+
+from . import cfp
+
+
+@cfp.cli.command("lottery")
+def lottery():
+    # In theory this can be extended to other types but currently only workshops & youthworkshops care
+    workshops = (
+        WorkshopProposal.query.filter_by(requires_ticket=True, type="workshop")
+        .filter(Proposal.state.in_(["accepted", "finalised"]))
+        .all()
+    )
+
+    app.logger.info(f"Running lottery for {len(workshops)} workshops")
+    winning_tickets = run_lottery(workshops)
+    app.logger.info(f"{len(winning_tickets)} won")
+
+    youthworkshops = (
+        WorkshopProposal.query.filter_by(requires_ticket=True, type="youthworkshop")
+        .filter(Proposal.state.in_(["accepted", "finalised"]))
+        .all()
+    )
+    app.logger.info(f"Running lottery for {len(youthworkshops)} youthworkshops")
+    winning_tickets = run_lottery(youthworkshops)
+    app.logger.info(f"{len(winning_tickets)} won")
+
+
+def run_lottery(ticketed_proposals):
+    """
+    Here are the rules for the lottery.
+    * Each user can only have one lottery ticket per workshop
+    * A user's lottery tickets are ranked by preference
+    * Drawings are done by rank
+    * Once a user wins a lottery their other tickets are cancelled
+    """
+    # Copy because we don't want to change the original
+    ticketed_proposals = ticketed_proposals.copy()
+    lottery_round = 0
+
+    winning_tickets = []
+
+    app.logger.info(f"Found {len(ticketed_proposals)} proposals to run a lottery for")
+    # Lock the lottery
+    signup = SiteState.query.get("signup_state")
+    if not signup:
+        raise Exception("'signup_state' not found.")
+
+    # This is the only state for running the lottery
+    signup.state = "run-lottery"
+    db.session.commit()
+    db.session.flush()
+    refresh_states()
+
+
+    max_rank = db.session.query(func.max(EventTicket.rank)).scalar() + 1
+    # Cache this locally so we're not hitting the db and not having to flush() etc.
+    proposal_capacities = {p.id: p.get_lottery_capacity() for p in ticketed_proposals}
+    winning_tickets = []
+
+    for lottery_round in range(max_rank):
+        for proposal in ticketed_proposals:
+            tickets_remaining = proposal_capacities[proposal.id]
+
+            tickets_for_round = [t for t in proposal.tickets if t.is_in_lottery_round(lottery_round)]
+            shuffle(tickets_for_round)
+
+            if tickets_remaining <= 0:
+                for ticket in tickets_for_round:
+                    ticket.lost_lottery()
+                    db.session.commit()
+
+                continue
+
+            for ticket in tickets_for_round:
+                if ticket.ticket_count < tickets_remaining:
+                    ticket.won_lottery_and_cancel_others()
+                    winning_tickets.append(ticket)
+                    tickets_remaining -= ticket.ticket_count
+                else:
+                    ticket.lost_lottery()
+                db.session.commit()
```

Comment:
I think it'll be better to not call `.commit` in either of these places. It's preferable to have all these changes go through in a single transaction, so they either all work or none of them. As a bonus we get a single transaction in the `transaction` table that ties all the changes together.

There's still the potential for some inconsistency if someone's resubmitting their tickets all the time, and the locking we gain from that actually helps prevent raciness.

Reply:
The major effect here (other than a likely big speedup) is that we eventually end up loading all active tickets into memory. That might be quite a bit, so I want to run a sanity check on it.